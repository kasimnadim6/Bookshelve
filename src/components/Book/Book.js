import React from 'react';
import styles from './Book.module.scss';

const Book = (props) => {
  const { book, inReadingList, updateReadingList } = props;
  return (
    <figure className={styles.book}>
      <img className={styles.image} src={book?.cover_image} alt={book?.title} />
      <figcaption data-testid="book-title">{book?.title}</figcaption>
      <button
        data-testid="btn-reading-status"
        className={`${styles['btn']} ${
          inReadingList ? styles.btn__add : styles.btn__remove
        }`}
        onClick={() => updateReadingList(book, inReadingList)}
      >
        {!inReadingList ? 'Add' : 'Remove'}
      </button>
    </figure>
  );
};

export default Book;
