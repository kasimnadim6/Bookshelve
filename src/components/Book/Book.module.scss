.book {
  margin: 0.1rem 0.6rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  font-size: 1rem;
  box-shadow: 1px 2px 3px rgba(0, 0, 0, 0.2);
  border-radius: 4px;
  &:hover {
    cursor: pointer;
    box-shadow: 3px 4px 7px rgba(0, 0, 0, 0.2);
    transform: translateY(-1px);
    transition: all 0.6s;
  }
  img {
    width: 12rem;
    height: 10rem;
  }
  figcaption {
    font-weight: 100;
    padding: 0.6rem;
  }
  .btn {
    font-family: inherit;
    text-transform: uppercase;
    outline: none;
    border: none;
    cursor: pointer;
    margin-bottom: 0.8rem;
    background-color: transparent;
    font-weight: bold;
    &__add {
      color: seagreen;
    }
    &__remove {
      color: darkred;
    }
  }
}
