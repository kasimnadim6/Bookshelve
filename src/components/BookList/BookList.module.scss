.book-list-container {
  display: flex;
  flex-direction: column;
  padding: 1rem;
  .header {
    text-align: center;
    padding: 1rem;
    font-size: 2.4rem;
  }
  .title {
    padding: 0.6rem;
    font-size: 1.2rem;
    color: salmon;
  }
  .book-list {
    display: flex;
    overflow-x: auto;
  }
  .readingList {
    margin-top: 2rem;
  }
}
