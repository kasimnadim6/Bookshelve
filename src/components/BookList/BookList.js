import { Component } from 'react';
import Book from '../Book/Book';
import styles from './BookList.module.scss';
import { Books } from '../../data';

export default class BookList extends Component {
  constructor() {
    super();
    this.state = {
      books: [],
      readingList: [],
    };
    this.updateReadingList = this.updateReadingList.bind(this);
    this.bookExistInReadingList = this.bookExistInReadingList.bind(this);
  }
  componentDidMount() {
    this.setState({
      books: Books,
    });
  }
  updateReadingList(book, inReadingList) {
    if (!inReadingList) {
      this.setState({
        readingList: [...this.state.readingList, book],
      });
    } else {
      this.setState({
        readingList: this.state.readingList.filter((f) => f.id !== book.id),
      });
    }
  }
  bookExistInReadingList(book) {
    return !!this.state.readingList.find((b) => b.id === book.id);
  }
  render() {
    const { books, readingList } = this.state;
    return (
      <section className={styles['book-list-container']}>
        <h1 data-testid="header" className={styles.header}>
          Welcome to My Favorite Book Portal
        </h1>
        <h2 className={styles.title}>Book Shelve</h2>
        <div className={styles['book-list']}>
          {books.length &&
            books.map((book) => (
              <Book
                key={book.id}
                book={book}
                inReadingList={this.bookExistInReadingList(book)}
                updateReadingList={this.updateReadingList}
              />
            ))}
        </div>
        {readingList.length ? (
          <div className={styles.readingList}>
            <h2 data-testid="reading-list" className={styles.title}>
              Reading List
            </h2>
            <div className={styles['book-list']}>
              {readingList.map((book) => (
                <Book
                  key={book.id}
                  book={book}
                  inReadingList={this.bookExistInReadingList(book)}
                  updateReadingList={this.updateReadingList}
                />
              ))}
            </div>
          </div>
        ) : (
          <p
            data-testid="no-reading-list"
            className={styles.title}
            style={{ textAlign: 'center', marginTop: '5rem' }}
          >
            Reading List Is Empty
          </p>
        )}
      </section>
    );
  }
}
