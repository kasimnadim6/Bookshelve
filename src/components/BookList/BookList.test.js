import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Book from '../Book/Book';
import BookList from './BookList';

const mockBookComponent = jest.fn();
jest.mock('../Book/Book', () => (props) => {
  mockBookComponent(props);
  return <mock-bookComponent />;
});
describe('BookList Component', () => {
  it('should render BookList component', () => {
    render(<BookList />);
    expect(BookList).toBeTruthy();
  });
  it(`should have text 'Welcome to My Favorite Book Portal'`, () => {
    render(<BookList />);
    expect(screen.getByTestId('header')).toHaveTextContent(
      'Welcome to My Favorite Book Portal'
    );
  });
  it('should have right BookList Snapshot', () => {
    const snapshotJSON = renderer.create(<BookList />).toJSON();
    expect(snapshotJSON).toMatchSnapshot();
  });
  it(`should have text 'Reading List Is Empty'`, () => {
    render(<BookList />);
    expect(screen.getByTestId('no-reading-list')).toHaveTextContent(
      'Reading List Is Empty'
    );
  });

  it(`should have 'Reading List'`, () => {
    const mock = [
      {
        id: 1,
        author_id: 4,
        title: 'Norwegian Wood',
        cover_image:
          'http://t1.gstatic.com/images?q=tbn:ANd9GcQvJJDi2mzwg9v_PlmKYL31gXIz0kvAObJak7DVFPcD_mJTIyec',
        pages: 296,
        releaseDate: '1987',
        isbn: 'hj846',
      },
    ];
    const updateReadingListMocked = jest.fn();
    render(<BookList />);
    expect(mockBookComponent).toHaveBeenCalled();
    // expect(mockBookComponent).toHaveBeenCalledWith(
    //   expect.objectContaining({
    //     book: mock,
    //     inReadingList: false,
    //     updateReadingList: updateReadingListMocked,
    //   })
    // );
  });
});
