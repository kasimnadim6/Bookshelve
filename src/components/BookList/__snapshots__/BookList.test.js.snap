// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList Component should have right BookList Snapshot 1`] = `
<section
  className="book-list-container"
>
  <h1
    className="header"
    data-testid="header"
  >
    Welcome to My Favorite Book Portal
  </h1>
  <h2
    className="title"
  >
    Book Shelve
  </h2>
  <div
    className="book-list"
  >
    <mock-bookComponent />
    <mock-bookComponent />
    <mock-bookComponent />
    <mock-bookComponent />
    <mock-bookComponent />
    <mock-bookComponent />
    <mock-bookComponent />
    <mock-bookComponent />
  </div>
  <p
    className="title"
    data-testid="no-reading-list"
    style={
      Object {
        "marginTop": "5rem",
        "textAlign": "center",
      }
    }
  >
    Reading List Is Empty
  </p>
</section>
`;
